### PR TITLE
Adding `abi_stable` support for `tremor-value`, v2

### PR DESCRIPTION
**Pull request by marioortizmanero**
# Pull request

## Description

This is an intermediate PR for #1287 so that it's more manageable. This is the second attempt for `tremor-value`, see #1303 for the first one. My last comment points out why I'm doing this:

> After talking with @darach, we both agreed that this might not be the best way to do it. Instead of having a single `Value` that's now `abi_stable` compatible, I suggest having two of them: one for the PDK, and another for regular usage. We can then convert to and from the regular type into the PDK type in order to communicate with the plugins.
> 
> * It's much simpler: `abi_stable` was propagating throughout the *entire* Tremor codebase and it was quickly getting out of hands. I was having trouble trying to make `tremor-script` compile, because it also uses `Value` for reasons outside of the PDK. I basically had to rewrite lots of stuff just for the PDK.
> * Less importantly: performance. At first it seems like it might be more costly, because having two types means there's a conversion from the regular value to the PDK value in order to pass it to the plugin, and then the inverse to use it in the plugin again. But it turns out that by using a single `Value` we were missing out on lots of optimizations, such as `known_key`, `raw_entry`, the memory efficiency of `beef::Cow`, or the performance of `halfbrown::HashMap`. And we were performing lots of conversions as well anyway, so it's not like this was better.
> 
> So yeah, I'm not fully sure having two value types is going to be better, but I'm certain it's worth trying. @Licenser was absolutely right that I shouldn't worry much about this PR because I was probably going to end up rewriting most of it anyway haha. At least having `value-trait` be `abi_stable`-compatible is still useful because I'll use it for the PDK value.
> 
> I'll try to rewrite this from almost scratch and let you know when I'm more or less done :)

In this PR, all I do is make `tremor-value` compatible with `abi_stable`. The main crate will still compile because it's a backward-compatible change. I'll soon make another PR with the changes to `tremor-runtime`, which are much more complicated than this simple PR.

Once this is approved and more or less ready, I'll merge https://github.com/marioortizmanero/tremor-runtime/pull/3 over it with the full PDK changes.

## Related

* [RFC](https://www.tremor.rs/rfc/accepted/plugin-development-kit)
* Related Issues: #791 #1077
* Related docs: https://nullderef.com/series/rust-plugins/

## Checklist

* [x] The RFC, if required, has been submitted and approved
* [ ] Any user-facing impact of the changes is reflected in docs.tremor.rs
* [ ] The code is tested
* [ ] Use of unsafe code is reasoned about in a comment
* [ ] Update CHANGELOG.md appropriately, recording any changes, bug fixes, or other observable changes in behavior
* [ ] The performance impact of the change is measured (see below)

### CODE REVIEW

File: src/pdk/value.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+pub type Object<'value> = RHashMap<RCow<'value, str>, Value<'value>>;
+/// Bytes
+pub type Bytes<'value> = RCow<'value, [u8]>;
+
+/// FFI-safe `Value` type to communicate with the plugins. It's meant to be
+/// converted to/from the original `tremor_value::Value` type and back so that
+/// it can be passed through the plugin interface.
+#[repr(C)]
+#[derive(StableAbi)]
+pub enum Value<'value> {
```
Comment (mfelsche) [inline src/pdk/value.rs:17]: If this is the `Value` PDK developers get into their hands i think we need to implement more traits from `value_trait` on this one, to make it usable. But this is not a concern for this PR. Please add a Note about it at least.

Comment (marioortizmanero) [reply src/pdk/value.rs:17]: Supposedly it's only used as an intermediate. The regular value is converted to the FFI one, and the in the plugin it's converted back to the regular one.

I still have to figure out how to do these conversions in the least annoying way for the plugin developer but it's definitely doable.

Comment (mfelsche) [reply src/pdk/value.rs:17]: Got it, i wasn't aware of how this was handled on the plugin side. 👍

**Comment by marioortizmanero** [comment]
Closing in favor of #1336

**Pull request closed without merging**